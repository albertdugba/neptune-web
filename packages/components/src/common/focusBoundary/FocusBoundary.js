import React, { useEffect, useRef } from 'react';
import Types from 'prop-types';

import { isKey, isUndefined } from '@transferwise/neptune-validation';

import { useConditionalListener } from '../hooks';
import { getFocusableElements, resetFocus } from './utils';
import { Key } from '../key';

const { TAB } = Key;

const FocusBoundary = ({ children }) => {
  const boundaryRef = useRef();
  const parent = isUndefined(document) ? undefined : document;
  let focusableEls = {};

  useEffect(() => {
    boundaryRef.current.focus();
    focusableEls = getFocusableElements(boundaryRef.current);
  }, []);

  // If event type is Tab the resetFocus will force the focus to either the first focusable or last in boundaryRef .
  useConditionalListener({
    eventType: 'keydown',
    callback: (event) => isKey({ keyType: TAB, event }) && resetFocus({ event, focusableEls }),
    attachListener: true,
    parent,
  });

  return (
    <span ref={boundaryRef} tabIndex={-1} className="np-focus-boundary d-inline-flex outline-none">
      {children}
    </span>
  );
};

FocusBoundary.propTypes = {
  children: Types.node.isRequired,
};

export default FocusBoundary;
