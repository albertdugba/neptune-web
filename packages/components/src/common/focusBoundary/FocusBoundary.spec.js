import React from 'react';
import { render, screen, userEvent, fireEvent } from '../../test-utils';
import FocusBoundary from './FocusBoundary';

const props = {
  onClose: jest.fn(),
};

describe('FocusBoundary', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('renders component', () => {
    const { container } = render(
      <FocusBoundary {...props}>
        <a href="test">Test</a>
      </FocusBoundary>,
    );

    expect(container).toMatchSnapshot();
  });

  it('focus on container by default', () => {
    render(
      <>
        <a href="test">not focusable</a>
        <FocusBoundary {...props}>
          <a href="test">Test</a>
        </FocusBoundary>
      </>,
    );

    expect(focusBoundary()).toHaveFocus();
  });

  it('traps the focus when user presses tab', () => {
    render(
      <>
        <a href="test">not focusable</a>
        <FocusBoundary {...props}>
          <a href="test">firstFocusable</a>
          <a href="test">lastFocusable</a>
        </FocusBoundary>
      </>,
    );

    userEvent.tab();
    expect(firstFocusableElement()).toHaveFocus();

    userEvent.tab();
    expect(lastFocusableElement()).toHaveFocus();

    userEvent.tab();
    expect(firstFocusableElement()).toHaveFocus();

    userEvent.tab({ shift: true });
    expect(lastFocusableElement()).toHaveFocus();

    userEvent.tab({ shift: true });
    expect(firstFocusableElement()).toHaveFocus();
  });

  it('calls onClose when user press Escape', () => {
    render(
      <FocusBoundary {...props}>
        <a href="test">firstFocusable</a>
        <a href="test">lastFocusable</a>
      </FocusBoundary>,
    );

    fireEvent.keyDown(document, { key: 'Escape' });
    expect(props.onClose).toHaveBeenCalled();
  });

  it(`doesn't call onClose when calls onClose when user press unsupported key`, () => {
    render(
      <FocusBoundary {...props}>
        <a href="test">firstFocusable</a>
        <a href="test">lastFocusable</a>
      </FocusBoundary>,
    );

    fireEvent.keyDown(document, { key: 'Enter' });
    expect(props.onClose).not.toHaveBeenCalled();
  });

  it('calls onClose when user clicks outside', () => {
    render(
      <FocusBoundary {...props}>
        <a href="test">firstFocusable</a>
        <a href="test">lastFocusable</a>
      </FocusBoundary>,
    );

    fireEvent.click(document);
    expect(props.onClose).toHaveBeenCalled();
  });

  it(`doesn't call onClose when user clicks in target`, () => {
    render(
      <FocusBoundary {...props}>
        <a href="test">firstFocusable</a>
        <a href="test">lastFocusable</a>
      </FocusBoundary>,
    );

    fireEvent.click(firstFocusableElement());
    expect(props.onClose).not.toHaveBeenCalled();
  });

  const firstFocusableElement = () => screen.getByText('firstFocusable');
  const lastFocusableElement = () => screen.getByText('lastFocusable');
  const focusBoundary = () => document.querySelector('.np-focus-boundary');
});
