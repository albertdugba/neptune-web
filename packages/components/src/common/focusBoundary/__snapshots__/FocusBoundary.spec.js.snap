// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusBoundary renders component 1`] = `
<div>
  <span
    class="np-focus-boundary outline-none"
    tabindex="-1"
  >
    <a
      href="test"
    >
      Test
    </a>
  </span>
</div>
`;
